src/lib/crypto/kimchi_backend/gadgets/bitwise.ml

```diff
@@ -141,8 +135,66 @@ let rot64 (type f)
 
   (* Next row *)
   Range_check.bits64 (module Circuit) shifted ;
+
+  (* Following row *)
+  Range_check.bits64 (module Circuit) excess ;
+
+  if check64 then Range_check.bits64 (module Circuit) word ;
+
+  (rotated, excess, shifted)
+
+(* 64-bit Rotation of rot_bits to the `mode` side
+ *   Inputs
+ *     - check: whether to check the input word is at most 64 bits (default is false)
+ *     - word of maximum 64 bits to be rotated
+ *     - rot_bits: number of bits to be rotated
+ *     - mode: Left or Right
+ * Output: rotated word
+ *)
+let rot64 (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    ?(check64 : bool = false) (word : Circuit.Field.t) (rot_bits : int)
+    (mode : rot_mode) : Circuit.Field.t =
+  let rotated, _excess, _shifted =
+    rot_aux (module Circuit) ~check64 word rot_bits mode
+  in
+
   rotated
 
+(* 64-bit bitwise logical shift of bits to the left side
+ * Inputs
+ *  - check64: whether to check the input word is at most 64 bits (default is false)
+ *  - word of maximum 64 bits to be shifted
+ *  - bits: number of bits to be shifted
+ * Output: left shifted word (with bits 0s at the least significant positions)
+ *)
+let lsl64 (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    ?(check64 : bool = false) (word : Circuit.Field.t) (bits : int) :
+    Circuit.Field.t =
+  let _rotated, _excess, shifted =
+    rot_aux (module Circuit) ~check64 word bits Left
+  in
```

Comment:
While I understand the attraction of generality, how much extra cost is it to compute the rotation when we just want the shift (left or right)? 

Or have we decided that it's okay to do it that way? Looks like we do not have dedicated circuitry for shifts

Reply:
We do not have dedicated circuitry for shifts nor want to have more custom gates for them if possible. Then, since rotation computed this way is basically equivalent to shifting and an addition, it seems like a very reasonable cost to support 2 more functionalities without new gates.